lib/std/crypto/25519/edwards25519.zig

```diff
@@ -0,0 +1,222 @@
+    /// Add two Edwards25519 points.
+    pub inline fn add(p: Edwards25519, q: Edwards25519) Edwards25519 {
+        const a = p.y.sub(p.x).mul(q.y.sub(q.x));
+        const b = p.x.add(p.y).mul(q.x.add(q.y));
+        const c = p.t.mul(q.t).mul(Fe.edwards25519d2);
+        var d = p.z.mul(q.z);
+        d = d.add(d);
+        const x = b.sub(a);
+        const y = b.add(a);
+        const z = d.add(c);
+        const t = d.sub(c);
+        return .{
+            .x = x.mul(t),
+            .y = y.mul(z),
+            .z = z.mul(t),
+            .t = x.mul(y),
+        };
+    }
+
+    inline fn cMov(p: *Edwards25519, a: Edwards25519, c: u64) void {
+        p.x.cMov(a.x, c);
+        p.y.cMov(a.y, c);
+        p.z.cMov(a.z, c);
+        p.t.cMov(a.t, c);
+    }
+
+    inline fn pcSelect(pc: [16]Edwards25519, b: u8) Edwards25519 {
+        var t = Edwards25519.identityElement();
+        comptime var i: u8 = 0;
+        inline while (i < 16) : (i += 1) {
+            t.cMov(pc[i], ((@as(usize, (b ^ i)) -% 1) >> 8) & 1);
+        }
+        return t;
+    }
+
+    fn pcMul(pc: [16]Edwards25519, s: [32]u8) !Edwards25519 {
+        var q = Edwards25519.identityElement();
+        var pos: usize = 252;
+        while (true) : (pos -= 4) {
+            q = q.dbl().dbl().dbl().dbl();
+            const b = (s[pos / 8] >> @intCast(u3, pos & 7)) & 0xf;
+            q = q.add(pcSelect(pc, b));
+            if (pos == 0) break;
+        }
+        try q.rejectIdentity();
+        return q;
+    }
+
+    fn precompute(p: Edwards25519) [16]Edwards25519 {
+        var pc: [16]Edwards25519 = undefined;
+        pc[0] = Edwards25519.identityElement();
+        pc[1] = p;
+        var i: usize = 2;
+        while (i < 16) : (i += 1) {
+            pc[i] = pc[i - 1].add(p);
+        }
+        return pc;
+    }
+
+    fn _mul(p: Edwards25519, s: [32]u8) !Edwards25519 {
+        var pc: [16]Edwards25519 = undefined;
+        if (p.is_base) {
+            @setEvalBranchQuota(10000);
+            pc = comptime precompute(Edwards25519.basePoint());
+        } else {
+            pc = precompute(p);
+            pc[4].rejectIdentity() catch |_| return error.WeakPublicKey;
+        }
+        return pcMul(pc, s);
+    }
+
+    /// Multiply an Edwards25519 point by a scalar after "clamping" it.
+    /// Clamping forces the scalar to be a multiple of the cofactor in
+    /// order to prevent small subgroups attacks.
+    /// This is strongly recommended for DH operations.
+    /// Return error.WeakPublicKey if the resulting point is
+    /// the identity element.
+    pub fn clampedMul(p: Edwards25519, s: [32]u8) !Edwards25519 {
+        var t: [32]u8 = s;
+        scalar.clamp(&t);
+        return _mul(p, t);
+    }
+
+    /// Multiply an Edwards25519 point by a scalar without clamping it.
+    /// Return error.WeakPublicKey if the resulting point is
+    /// the identity element.
+    pub fn mul(p: Edwards25519, s: [32]u8) !Edwards25519 {
+        return _mul(p, s);
```

Comment:
Why is this not the actual implementation of `_mul`?

Reply:
Done.